### DISCUSSION TITLE
Plotting after arcsinh transformation and using multiple cofactors

### DISCUSSION THREAD

**Discussion by phuongtran807** [ImmuneDynamics/Spectre#176] [Help]
Hi!
First of all, thanks for the great package and documentation.
I am currently following the instruction 'Discovery workflow with batch alignment using CytoNorm' and was wondering how to procede if I'm choosing different cofactors for different channels. Also, I have a problem with plotting after the arcsinh transformation.

for(i in transformed.cols){
  make.colour.plot(do.subsample(cell.dat, 20000), i, plot.against)
}

It runs about half an hour and then I get the warning message ('In grDevices::dev.off() : agg could not write to the given file') and no output.
It does give me an output when I'm doing it without the for-loop and by defining one specific marker tho but that would take some time with 16 different markers.. 

Thanks for your help!

Phuong

**Comment by csecarroll** [comment]
yeah. I've found make.colour.plot is always a problem. After running it (and the heatmap too) I have issues with all my plots. Not make.multi.plot for some reason.  I cant seem to get it to print in a markdown document. I even have issues inserting images after running it. 

My work around was to create a function to automatically test a range of cofactors and use ggplot to create facet plots for all my markers (I'm using 20 markers), with one facet plot per co-factor. Then pick the appropriate co-factor for the data, and apply it to the dataset.  

If anyone finds a way to fix the make.colour.plot issue, id be very interested.

**Comment by tomashhurst** [comment]
@phuongtran807 @csecarroll thanks for reaching out! I don't get the same issue, but I have a feeling I know what the issue is the `dev.off()` is the clue, a classic R thing that messes with plots. 

As a first pass, let's try a quick test: could you try adding `dev.off()` after the `make.colour.plot()` function inside the loop. If it runs OK, then this suggests that the end of the plotting in `make.colour.plot` is causing the headache. In general you can run `dev.off()` after any plotting and it will help clean things up, may even solve the heatmap issues.

In terms of specifying different co-factors, you can essentially just run `do.asinh` twice -- each time on a different set of columns, and with a different co-factor. Here is an example:

```
## Setup
library('Spectre')
dat <- Spectre::demo.clustered
dat <- dat[,1:10]
dat <- do.subsample(dat, 10000)
dat

## Set cofactors 
as.matrix(names(dat))

low.asinh <- names(dat)[c(2:5)] 
high.asinh <- names(dat)[c(6:10)]

low.cf <- 500
high.cf <- 5000

## Apply cofactors

dat <- do.asinh(dat = dat, use.cols = low.asinh, cofactor = low.cf)
dat <- do.asinh(dat = dat, use.cols = high.asinh, cofactor = high.cf)

## Review
dat
```

Note in this case I have just picked the co-factors and channels at random, not saying these choices would be correct for this dataset.

**Comment by csecarroll** [comment]
Hello! Thanks for the quick response (and for the great package). 

So I've tried a few things out this morning. 

If I'm using make.colour.plot to plot say FSC and SSC, I have no issues. It outputs fine and I can also output a ggplot version afterwards with no issues. 

However,  after using make.colour.plot to do a UMAP or tSNE plot, I no longer get the output (it opens the plot panel but it remains blank), and then I can no longer get a ggplot plot to output anymore either.  

if i add dev.off() it does seem to fix the problem.  I was adding graphics.off() previously, which didn't fix it.  They now show up in the plots pane and I can run a ggplot plot afterwards with no issues. 

I've been using Spectre almost daily for months now and its great to have these working. 

However, with the heatmaps, dev.off() doesn't seem to fix the issues with them. I seem to have an issue assigning them to an object too. 

for example 
  hm1 <-   make.pheatmap(exp, "FlowSOM_metacluster", cellular.cols, path = clust.dir, file.name = heatmap.png")

>A pheatmap has been saved to your working directory

hm1
> hm1
NULL

adding dev.off() did not fix it. 

Thanks for your help

**Comment by tomashhurst** [comment]
@csecarroll now that is weird. Could you try plotting with the clustered demo dataset? 

```
library('Spectre')
dat <- Spectre::demo.clustered
make.colour.plot(dat, 'UMAP_X', 'UMAP_Y', 'FlowSOM_metacluster')
```

just as a sanity check for the heatmap, but for the filename you might be missing quotations? `file.name = heatmap.png")` should be `file.name = "heatmap.png")`?

**Comment by csecarroll** [comment]
yes. I've checked and i had the correct quotations (I had just removed a very specific part of the name in the github comment but it was correct in R). 

I also don't have any issues if I run the normal pheatmap function on exp. 

I'm not sure if it could be an issue with using vs code instead of R studio. I can try running it in Rstudio and see what happens. When I get some time, I'll give the demo dataset a go. 

thanks again.

**Comment by ghar1821** [comment]
@csecarroll I think I managed to replicate your problem on vs code. 

So after running `make.pheatmap`, I could not get `make.colour.plot` to produce any plot at all. The function will run, but the plot won't show up in the plot panel and it won't be saved in the directory I specified either. If I ran `dev.off()` after running `make.pheatmap`,  `make.colour.plot` will then run as per normal. I'll need to dig into the pheatmap code to see what is going on. 

Also, the `make.pheatmap` function is not meant to return any object, which is why your `hm1` object is NULL. This is expected. I don't think it is possible to get a ggplot object (or other type of object) out of pheatmap unfortunately. However, we can change the function so it shows the plot instead of always saving it in a file.
